---
title: "R Notebook"
output:
  html_notebook: Methane generation and methanogenic community structure in an anaerobic hybrid bioreactor for biologically enhanced primary treatment of domestic wastewater under ambient temperatures
  html_document: Methane generation and methanogenic community structure in an anaerobic hybrid bioreactor for biologically enhanced primary treatment of domestic wastewater under ambient temperatures
---
```{r message=FALSE, warning=FALSE}
# check.packages function: install and load multiple R packages.
# Check to see if packages are installed. Install them if they are not, then load them into the R session.
check.cran <- function(pkg){
    new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
    if (length(new.pkg)) 
        install.packages(new.pkg, dependencies = TRUE)
    sapply(pkg, require, character.only = TRUE)
}
packages.cran<-c("ape", "tidyverse", "here", "devtools","remotes")
check.cran(packages.cran)

source("http://bioconductor.org/biocLite.R")
check.bioconductor <- function(pkg){
    new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
    if (length(new.pkg)) 
        biocLite(new.pkg, dependencies = TRUE, suppressUpdates = TRUE)
    sapply(pkg, require, character.only = TRUE)
}
packages.bioconductor<-c("phyloseq", "ShortRead", "dada2", "metagenomeSeq")
check.bioconductor(packages.bioconductor)

#remotes::install_github("MadsAlbertsen/ampvis2")
library(ampvis2)
```


```{r}
paste0("phyloseq version ",packageVersion("phyloseq"))
paste0("ShortRead version ",packageVersion("ShortRead"))
paste0("dada2 version ",packageVersion("dada2"))
paste0("ape version ",packageVersion("ape"))
paste0("tidyverse version ",packageVersion("tidyverse"))
paste0("metagenomeSeq version ",packageVersion("metagenomeSeq"))
paste0("ampvis2 version ",packageVersion("ampvis2"))
```


###This workflow also requires Qiime 1.9.1 and sra-tools installed using Anaconda or miniconda with Python 2.7.

```{r}
library(here)
#This is where things will go just in case you are looking for them
here()
```

```{r Code you need to chage}
#YOU WILL HAVE TO CHANGE THIS TO THE APPROPRIATE DIRECTORY!!!!!
#this allows R to call qiime 1.9 and sra-tools. They both need to be installed in the python2 environment.  If you have anaconda3 and create an environment called python2 that's running python 2.7 and has QIIME and sra-tools installed the following might work.
Sys.setenv(PATH = paste("/anaconda3/envs/python2/bin/", Sys.getenv("PATH"),sep=":"))
```

```{r}
#we'll be using Mac's Bash, so let's make sure the working directory = here()
system(paste0("cd ", here()))
```

```{r}
#download the approriate SRA files
a<-c(paste0("SRR686940",1:9), paste0("SRR68694",10:13))
for (i in a){
     download.file(paste0("ftp://ftp-trace.ncbi.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR686/", i, "/", i, ".sra"),destfile=here(i))
 }
```

```{r}
#convert SRA files to fastq
system(paste0("cd ", here()))
for (i in a) {
system(paste0("fastq-dump ",i," --split-3") )
}
```

```{r}
#make directories for forward and reverse fastq files
dir.create('fastqsF')
dir.create('fastqsR')
```

```{r}
#copy forward fastqs to fastqF directory, and reverse fastqs to the fastqR directory
list.of.forward.fastqs <- list.files(here(), "_1.fastq")
list.of.reverse.fastqs <- list.files(here(), "_2.fastq")
 # copy the files to the new folder
file.copy(list.of.forward.fastqs, paste0(here("fastqsF")))
file.copy(list.of.reverse.fastqs, paste0(here("fastqsR")))

```

```{r}
#start the DADA2 workflow with filterAndTrim
pathF <- here("fastqsF")
pathR <- here("fastqsR")
filtpathF <- file.path(pathF, "filtered") # Filtered forward files go into the pathF/filtered/ subdirectory
filtpathR <- file.path(pathR, "filtered") # ...
fastqFs <- sort(list.files(pathF, pattern=".fastq"))
fastqRs <- sort(list.files(pathR, pattern=".fastq"))
if(length(fastqFs) != length(fastqRs)) stop("Forward and reverse files do not match.")
track.filt <-filterAndTrim(fwd=file.path(pathF, fastqFs), filt=file.path(filtpathF, fastqFs),
              rev=file.path(pathR, fastqRs), filt.rev=file.path(filtpathR, fastqRs),
              trimLeft=c(40, 20), maxEE=2, truncQ=2, maxN=0, rm.phix=TRUE, compress=TRUE, verbose=TRUE, multithread=TRUE)

```

```{r}
# take filtered fastqs and return inferred sequences in a sequence table
filtFs <- list.files(filtpathF, pattern=".fastq", full.names = TRUE)
filtRs <- list.files(filtpathR, pattern=".fastq", full.names = TRUE)
sample.names <- sapply(strsplit(basename(filtFs), "_"), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz
sample.namesR <- sapply(strsplit(basename(filtRs), "_"), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz
if(!identical(sample.names, sample.namesR)) stop("Forward and reverse files do not match.")
names(filtFs) <- sample.names
names(filtRs) <- sample.names
set.seed(100)
# Learn forward error rates
errF <- learnErrors(filtFs, nread=2e6, multithread=TRUE)
# Learn reverse error rates
errR <- learnErrors(filtRs, nread=2e6, multithread=TRUE)
# Sample inference and merger of paired-end reads
mergers <- vector("list", length(sample.names))
ddFs <- vector("list", length(sample.names))
names(mergers) <- sample.names
names(ddFs) <- sample.names
for(sam in sample.names) {
  cat("Processing:", sam, "\n")
    derepF <- derepFastq(filtFs[[sam]])
    ddF <- dada(derepF, err=errF, multithread=TRUE)
    derepR <- derepFastq(filtRs[[sam]])
    ddR <- dada(derepR, err=errR, multithread=TRUE)
    merger <- mergePairs(ddF, derepF, ddR, derepR, minOverlap = 10)
    mergers[[sam]] <- merger
    ddFs[[sam]]<-ddF
}
rm(derepF); rm(derepR)
# Construct sequence table and remove chimeras
seqtab <- makeSequenceTable(mergers)
```

```{r}
#remove chimeras from the sequence table
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE)
```

```{r}
#look at sequence processing stats
getNreads <- function(x) sum(getUniques(x))
track <- cbind(track.filt, sapply(ddFs, getNreads), sapply(mergers, getNreads), rowSums(seqtab), rowSums(seqtab.nochim))
colnames(track) <- c("input", "denoised", "filtered", "merged","tabled", "nonchim")
track
```


```{bash, echo=TRUE}
#download silva training set for taxonomy assignment
file=silva_nr_v128_train_set.fa.gz
if [ ! -e "$file" ]; then
    wget -q https://zenodo.org/record/824551/files/silva_nr_v128_train_set.fa.gz

else 
    echo "silva_nr_v128_train_set.fa.gz exists, proceede to assignTaxonomy"
fi 
```
#download the approriate taxonomy databases
if (!file.exists(here("silva_database","silva_nr_v128_train_set.fa.gz"))
   download.file (https://zenodo.org/record/824551/files/silva_nr_v128_train_set.fa.gz, destfile=here("silva_database","silva_nr_v128_train_set.fa.gz")))
if (!file.exists(here("silva_database","silva_species_assignment_v128.fa.gz"))
   download.file (https://zenodo.org/record/824551/files/silva_species_assignment_v128.fa.gz, destfile=here("silva_database","silva_species_assignment_v128.fa.gz")))

```{r}
#assign taxonomy
tax <- assignTaxonomy(seqtab.nochim, "/Users/garyvanzin/Downloads/compressed_archives/silva_nr_v128_train_set.fa.gz", multithread=TRUE)
```

```{bash, echo=TRUE}
#download silva training set for species assignment
file=silva_species_assignment_v128.fa.gz
if [ ! -e "$file" ]; then
    wget -q https://zenodo.org/record/824551/files/silva_species_assignment_v128.fa.gz

else 
    echo "silva_species_assignment_v128.fa.gz exists, proceede to addSpecies"
fi 
```

```{r}
#assign species level taxonomy
genus.species <- addSpecies(tax, "/Users/garyvanzin/Downloads/compressed_archives/silva_species_assignment_v128.fa.gz")
```


```{r}
#import metadata
meta<-structure(list(Date = structure(c(17158, 17158, 17158, 16994, 
16994, 16994, 16994, 16994, 16994, 16994, 17158, 17158, 17158
), class = "Date"), `Day of Operation` = c(395L, 395L, 395L, 
231L, 231L, 231L, 231L, 231L, 231L, 231L, 395L, 395L, 395L), 
    `Location of Sample` = c("MP Cell 3", "MP Cell 1", "MP Cell 2", 
    "MP Influent", "MP Influent", "MP Cell 3", "MP Cell 2", "MP Cell 2", 
    "MP Cell 3", "MP Cell 1", "MP Cell 3", "MP Cell 2", "MP Influent"
    ), location = c("MP", "MP", "MP", "MP", "MP", "MP", "MP", 
    "MP", "MP", "MP", "MP", "MP", "MP"), cell = c("3", "1", "2", 
    "Influent", "Influent", "3", "2", "2", "3", "1", "3", "2", 
    "Influent"), season = c("winter", "winter", "winter", "summer", 
    "summer", "summer", "summer", "summer", "summer", "summer", 
    "winter", "winter", "winter")), .Names = c("Date", "Day of Operation", 
"Location of Sample", "location", "cell", "season"), row.names = c("SRR6869401", 
"SRR6869402", "SRR6869403", "SRR6869404", "SRR6869405", "SRR6869406", 
"SRR6869407", "SRR6869408", "SRR6869409", "SRR6869410", "SRR6869411", 
"SRR6869412", "SRR6869413"), spec = structure(list(cols = structure(list(
    order = structure(list(), class = c("collector_integer", 
    "collector")), SRA_number = structure(list(), class = c("collector_character", 
    "collector")), Date = structure(list(format = "%m/%d/%y"), .Names = "format", class = c("collector_date", 
    "collector")), `Day of Operation` = structure(list(), class = c("collector_integer", 
    "collector")), `Location of Sample` = structure(list(), class = c("collector_character", 
    "collector")), location = structure(list(), class = c("collector_character", 
    "collector")), cell = structure(list(), class = c("collector_character", 
    "collector")), season = structure(list(), class = c("collector_character", 
    "collector"))), .Names = c("order", "SRA_number", "Date", 
"Day of Operation", "Location of Sample", "location", "cell", 
"season")), default = structure(list(), class = c("collector_guess", 
"collector"))), .Names = c("cols", "default"), class = "col_spec"), class = c("tbl_df", 
"tbl", "data.frame"))
```

```{r, message=FALSE, warning=FALSE, include=FALSE}
#make the phyloseq object
ps_DI <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(meta), 
               tax_table(genus.species))

```

```{r}
#export unique sequences
a<-colnames(otu_table(ps_DI))
uniquesToFasta(seqtab.nochim, paste0(here(),"/ps_DI_MP.fasta"), ids = a)

```


```{r}
#align unique sequences with pynast
system('python /anaconda3/envs/python2/bin/parallel_align_seqs_pynast.py -i ps_DI_MP.fasta -o pynast_aligned_ps_DI_MP/ -T -O 4')
#filter alignment
system(paste0('python /anaconda3/envs/python2/bin/filter_alignment.py', ' -i ', here('/pynast_aligned_ps_DI_MP','ps_DI_MP_aligned.fasta'), ' -o filtered_alignment/'))
#make tree
system(paste0('python /anaconda3/envs/python2/bin/make_phylogeny.py', ' -i ', here('/filtered_alignment','ps_DI_MP_aligned_pfiltered.fasta')))

```

```{r}
#import tree back into phyloseq
tree = read.tree(here('/filtered_alignment','ps_DI_MP_aligned_pfiltered.tre'))
tree = root(tree, 1, resolve.root = T)
ps_DI_tree=merge_phyloseq(ps_DI,tree)
```


```{r}
#remove singletons, add in some metadata, normalize by cumulative sum scaling using the package metagenomeSeq
ps_DI_tree_MC2 <- filter_taxa(ps_DI_tree, function(x) max(x) >= 1, TRUE)
sample_data(ps_DI_tree_MC2)$temperature<-ifelse(sample_data(ps_DI_tree_MC2)$season=='summer','warm','cold')
metagenomeSeq_norm <- function(physeq) {
    # Enforce orientation
    if(!taxa_are_rows(physeq)){physeq <- t(physeq)}
    OTU <- as(otu_table(physeq), "matrix")
    # Convert sample_data to AnnotatedDataFrame
    ADF <- AnnotatedDataFrame(data.frame(sample_data(physeq)))
    # define dummy "feature" data for OTUs, using their name
    # Helps with extraction and relating to taxonomy later on.
    TDF <- AnnotatedDataFrame(
        data.frame(OTUname = taxa_names(physeq), row.names=taxa_names(physeq)))
    # Create the metagenomeSeq object
    MGS <- newMRexperiment(counts=OTU, phenoData=ADF, featureData=TDF)
    # Trigger metagenomeSeq to calculate its Cumulative Sum scaling factor.
    ptm <- proc.time()
    MGS <- cumNorm(MGS)
    scaleFac <- normFactors(MGS)
    normalizedCounts <- cumNormMat(MGS, sl = median(sample_sums(physeq)))
    runtime <- proc.time() - ptm
    return(list(normCounts = normalizedCounts, scaleFactor = scaleFac,
                runtime = runtime))
}
metaSeq_cumNorm <- metagenomeSeq_norm(ps_DI_tree_MC2)

ps_DI_tree_MC2_norm<-phyloseq(otu_table(metaSeq_cumNorm$normCounts,taxa_are_rows = TRUE), sample_data(sample_data(ps_DI_tree_MC2)),tax_table(tax_table(ps_DI_tree_MC2)),phy_tree(phy_tree(ps_DI_tree_MC2)))




```

```{r}
#add some metadata for plotting purposes
sample_data(ps_DI_tree_MC2_norm)$cell_season<-ifelse(sample_data(ps_DI_tree_MC2_norm)$cell=='Influent',paste0(sample_data(ps_DI_tree_MC2_norm)$cell," ",sample_data(ps_DI_tree_MC2_norm)$season),paste0("Cell ",sample_data(ps_DI_tree_MC2_norm)$cell," ",sample_data(ps_DI_tree_MC2_norm)$season))


```

```{r}
#perform ordination
ordu = ordinate(ps_DI_tree_MC2_norm, "PCoA", "unifrac", weighted=TRUE)
pc1_percent_explained <-paste0( "PCO1 ( ", round(ordu$values$Relative_eig[1]*100, 2),"% variance explained)")
pc2_percent_explained <-paste0( "PCO2 ( ", round(ordu$values$Relative_eig[2]*100, 2),"% variance explained)")
```


```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
#plot ordination
p<-plot_ordination(ps_DI_tree_MC2_norm, ordu, color = "temperature",  shape="cell")
#p<-p + geom_point(size=5, alpha=c(rep(.9,7),rep(.5,6)))#add points and alpha
p<-p + geom_point(size=5, alpha=c(rep(.5,3),rep(.9,7), rep(.5,3)))#add points and alpha
p<-p+scale_color_manual(values=c("blue","red"))#blue cold red warm
p<-p+scale_shape_manual(values=c(15,16,17,18),breaks=c("Influent","1","2","3") )#set shapes and order of legend
p<-p+labs(x=pc1_percent_explained, y=pc2_percent_explained)#set x and y axes
#p<-p+xlim(-0.26,0.46)+ylim(-0.26,0.46)#equal range
p<-p+xlim(-0.27,0.27)+ylim(-0.27,0.27)#square
```
```{r}
p + guides(colour = guide_legend(override.aes = list(alpha = c(0.5,1))))#makes legend look like figure
```


```{r}
#build phylum level heatmap backend data
obj <- ps_DI_tree_MC2
#obj <- ps_DI_tree_MC2_top50
#obj = subset_taxa(obj, Class!="NA")
otu_table(obj)<-t(otu_table(obj))
short_names <- substr(rownames(otu_table(obj)), 1, 5)%>%
 make.names(unique = TRUE)
taxa_names(obj) <- short_names
obj <- transform_sample_counts(obj, function(x) x/sum(x) * 100)
physeqrF = filter_taxa(obj, function(x) mean(x) < .1,TRUE)
rmtaxa = taxa_names(physeqrF)
alltaxa = taxa_names(obj)
myTaxa = alltaxa[!alltaxa %in% rmtaxa]
obj <- prune_taxa(myTaxa,obj)

otutable <- data.frame(OTU = rownames(phyloseq::otu_table(obj)@.Data),
                       phyloseq::otu_table(obj)@.Data,
                       phyloseq::tax_table(obj)@.Data,
                       check.names = FALSE
                       )


#Extract metadata from the phyloseq object:
metadata_ampvis <- data.frame(phyloseq::sample_data(obj), 
                       check.names = FALSE
                       )
metadata_ampvis=cbind(SampleID = rownames(metadata_ampvis), metadata_ampvis) 
# metadata$grouping<-paste0(metadata$Location.of.Sample," ", metadata$season)
# metadata$cell<-gsub("influent","Influent",metadata$cell)
# metadata$cell_season<-paste0("Cell ",metadata$cell," ", metadata$season)
#Load the data with amp_load:
my_ampvis2_object <- amp_load(otutable, metadata_ampvis)
#data to plot dominant phyla
subsettedData <- amp_subset_taxa(my_ampvis2_object, c('Euryarchaeota','Proteobacteria','Firmicutes','Bacteroidetes','Synergistetes','Aminicenantes','Spirochaetae'),normalise = TRUE)
#data to plot Euryarchaeota only
subsettedData_Euryarchaeota <- amp_subset_taxa(my_ampvis2_object, c('Euryarchaeota'),normalise = TRUE)

# subsettedData <- amp_subset_taxa(my_ampvis2_object,
#                                  aaa,
#                                  normalise = TRUE #add this
# )
# subsettedData$tax$Genus<-ifelse(subsettedData$tax$Genus=="",paste0("Family","_",subsettedData$tax$Family ,"_","Genus_NA"),subsettedData$tax$Genus)
# subsettedData$tax$Genus<-ifelse(subsettedData$tax$Genus=="Family__Genus_NA",paste0("Order", "_",subsettedData$tax$Order,"_" ,"Family_Genus_NA"),subsettedData$tax$Genus)

```

```{r}
#generate base heatmap, sort X axis
phylum_heatmap<-amp_heatmap(subsettedData,normalise = FALSE,group_by = "cell",plot_values_size=6, facet_by = 'temperature',color_vector = c("white", "darkred"))
phylum_heatmap$data$Group<-factor(phylum_heatmap$data$Group, levels=c('Influent','1','2','3'))


```
phylum_heatmap$data <- phylum_heatmap$data[order(as.numeric(as.character(phylum_heatmap$data$Abundance))), ]
```{r}
#cleanup fonts
phylum_heatmap<-phylum_heatmap + theme(axis.text.x = element_text(angle = 45, size=14, vjust = 1),
          axis.text.y = element_text(size=12), 
           legend.position="right", legend.text=element_text(size=12))
```

```{r message=FALSE, warning=FALSE, paged.print=FALSE}
plot(phylum_heatmap)
```


```{r}
#generate base euryarchaeota heatmap, sort X axis
euryarcheota_heatmap<-amp_heatmap(subsettedData_Euryarchaeota,normalise = FALSE,group_by = "cell", tax_aggregate = "Genus",tax_show = 4,round =1 ,plot_values_size=6, facet_by = 'temperature',color_vector = c("white", "darkred"))
euryarcheota_heatmap$data$Group<-factor(euryarcheota_heatmap$data$Group, levels=c('Influent','1','2','3'))

```

```{r}
#cleanup fonts
euryarcheota_heatmap<-euryarcheota_heatmap + theme(axis.text.x = element_text(angle = 45, size=14, vjust = 1),
          axis.text.y = element_text(size=12), 
           legend.position="right", legend.text=element_text(size=12))
```

```{r}
plot(euryarcheota_heatmap)
```

####supplementary figure
```{r}
#heatmap of top 40 taxa in the 7 dominant Phyla
ha_g<-amp_heatmap(subsettedData, normalise = FALSE,tax_empty = "best",
#ha_g<-amp_heatmap(my_ampvis2_object,
            group_by = "cell", facet_by = 'temperature',#,textmap=TRUE,
tax_aggregate = "Genus",tax_add = "Phylum",
            tax_show = 40,round =2 ,
            color_vector = c("white", "darkred"),
            #order_x_by = c('Influent','1','2','3' ),
            plot_colorscale = "sqrt", 
            plot_values = TRUE, plot_values_size = 2) + 
    theme(axis.text.x = element_text(angle = 45, size=10, vjust = 1),
          axis.text.y = element_text(size=10), 
           legend.position="right", legend.text=element_text(size=10))

ha_g$data$Group<-factor(ha_g$data$Group, levels=c('Influent','1','2','3'))
```

```{r}
#cleanup fonts
ha_g<-ha_g + theme(axis.text.x = element_text(angle = 45, size=14, vjust = 1),
          axis.text.y = element_text(size=12), 
           legend.position="right", legend.text=element_text(size=12))
```



```{r}
ha_g$data<-  ha_g$data[complete.cases( ha_g$data), ]
ha_g$data<-separate(ha_g$data, Display,into=c("Phylum","Genus"), sep=";",extra='drop', remove=FALSE)
ha_g$data$Genus<-gsub(" ","",ha_g$data$Genus)

```


```{r}
ha_g$data$Group<-factor(ha_g$data$Group, levels=c('Influent','1','2','3'))

test2<-ggplot(ha_g$data, aes((Group), Genus ),label = formatC("Abundance", format = "f", digits = 1)) +
  geom_tile(aes(fill = Abundance) , color = 'white') +
  scale_fill_gradient(low = "white", high = "darkred",trans="log10", na.value='white') +#default na.value is gray
  ylab("Genus ") +
  xlab("Location") +
  theme(legend.title = element_text(size = 10),
        legend.text = element_text(size = 12),
        plot.title = element_text(size=16),
        axis.title=element_text(size=14,face="bold"),
        axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(fill = "consortium percentage")
test2$data$Group<-factor(test2$data$Group, levels=c('Influent','1','2','3'))

```

```{r}
#test2 <- test2 + coord_equal()

test2<-test2+facet_grid(test2$data$Phylum~test2$data$temperature, scales='free', space='free')+theme(strip.text.y = element_text(angle = 0))#,strip.background = element_rect(colour="gray95", fill="gray95"))
test2<-test2+ geom_text(aes(label = round(Abundance, 1)))
```

```{r}
#re-order phylum
test2$data$Phylum<-factor(test2$data$Phylum,levels=c('Euryarchaeota','Proteobacteria','Firmicutes','Bacteroidetes','Synergistetes','Aminicenantes','Spirochaetae'))
```

```{r}
test2<-test2+ theme(strip.text.x = element_text(margin = margin(.25,0,.25,0, "cm"),size = 15))
test2$data$Genus<-gsub("_"," ", test2$data$Genus)
test2$data$Genus<-factor(test2$data$Genus)
test2$data$Genus<-fct_reorder(test2$data$Genus,test2$data$Abundance,fun=mean)
```

```{r}
test2<-test2+scale_x_discrete(expand=c(0,0)) + scale_y_discrete(expand=c(0,0))
```

```{r}
plot(test2)
```

```{r}
sessionInfo()
```





